lib/spack/spack/version.py

```diff
@@ -335,387 +366,198 @@ def up_to(self, index):
+        raise ValueError(f"Unexpected type {type(other)}")
 
-    def __repr__(self):
-        return "GitVersion(" + repr(self.string) + ")"
+    def overlaps(self, other) -> bool:
+        return self.intersects(other)
 
-    @coerced
-    def __lt__(self, other):
-        """Version comparison is designed for consistency with the way RPM
-        does things.  If you need more complicated versions in installed
-        packages, you should override your package's version string to
-        express it more sensibly.
-        """
-        if other is None:
+    def satisfies(
+        self, other: Union["GitVersion", StandardVersion, "ClosedOpenRange", "VersionList"]
+    ):
+        # Concrete versions mean we have to do an equality check
+        if isinstance(other, GitVersion):
+            return self == other
+        if isinstance(other, StandardVersion):
             return False
+        if isinstance(other, ClosedOpenRange):
+            return self.ref_version.satisfies(other)
+        if isinstance(other, VersionList):
+            return any(self.satisfies(rhs) for rhs in other)
+        raise ValueError(f"Unexpected type {type(other)}")
 
-        # If we haven't indexed yet, can't compare
-        # If we called this, we know at least one is a git ref
-        if not (self.ref_lookup or other.ref_lookup):
-            return False
+    def __str__(self):
+        s = f"git.{self.ref}" if self.has_git_prefix else self.ref
+        # Note: the solver actually depends on str(...) to produce the effective version.
+        # So when a lookup is attached, we require the resolved version to be printed.
+        # But for standalone git versions that don't have a repo attached, it would still
+        # be nice if we could print @<hash>.
+        if self._ref_version or self._ref_lookup:
+            s += f"={self.ref_version}"
+        return s
 
-        # Use tuple comparison assisted by VersionStrComponent for performance
-        return self._cmp(other.ref_lookup) < other._cmp(self.ref_lookup)
+    def __repr__(self):
+        return f'GitVersion("{self}")'
+
+    def __bool__(self):
+        return True
 
-    @coerced
     def __eq__(self, other):
-        # Cut out early if we don't have a git version
-        if other is None or type(other) != GitVersion:
-            return False
+        # GitVersion cannot be equal to StandardVersion, otherwise == is not transitive
+        return (
+            isinstance(other, GitVersion)
+            and self.ref == other.ref
+            and self.ref_version == other.ref_version
+        )
 
-        return self._cmp(other.ref_lookup) == other._cmp(self.ref_lookup)
+    def __ne__(self, other):
+        return not self == other
 
-    def __hash__(self):
-        return hash(str(self))
+    def __lt__(self, other):
+        if isinstance(other, GitVersion):
+            return (self.ref_version, self.ref) < (other.ref_version, other.ref)
```

Comment:
I guess this ultimately comes from: `a == b` should imply `hash(a) == hash(b)`. We don't want `hash(GitVersion)` to resolve `ref_version` since it should be possible to create dicts/sets of them when they're not yet resolved (package.versions is in fact such a dict). And `hash(GitVersion)` must be constant, so `ref` is the only hashable property. That means `==` cannot discard `ref`. So, `GitVersion.__eq__` makes sense if it's implmented in terms of `(ref, ref_version)`. And then I guess `<` is just using the same comparator. (It also ensures there's a total order, but idk if that's actually necessary.)